File: 2021/10/puzzle.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Stack []string

func (s *Stack) IsEmpty() bool {
	return len(*s) == 0
}

func (s *Stack) Push(str string) {
	*s = append(*s, str)
}

func (s *Stack) Pop() (string, bool) {
	if s.IsEmpty() {
		return "", false
	} else {
		index := len(*s) - 1
		element := (*s)[index]
		*s = (*s)[:index]
		return element, true
	}
}

func main() {
	fmt.Println("Day 10: Syntax Scoring")

	file, _ := os.Open("input.txt")
	defer file.Close()
	scanner := bufio.NewScanner(file)

	syntaxMap := map[string]string{"(": ")", "<": ">", "[": "]", "{": "}"}
	var errors []string
	for scanner.Scan() {
		line := strings.Split(scanner.Text(), "")
		var openers Stack
		for _, char := range line {
			if syntaxMap[char] != "" {
				openers.Push(char)
			} else {
				top, _ := openers.Pop()
				if syntaxMap[top] != char {
					errors = append(errors, char)
					break
				}
			}
		}
	}

	pointMap := map[string]int{")": 3, ">": 25137, "]": 57, "}": 1197}
	score := 0
	for _, err := range errors {
		score += pointMap[err]
	}

	fmt.Println(fmt.Sprintf("Part One: %d", score))
}

File: 2022/1/puzzle.py
print("Day 1: Calorie Counting")

with open("input.txt", encoding="utf-8") as f:
    snacks = list(f)

calories = [0]
i = 0
for snack in snacks:
    if snack == "\n":
        i += 1
        calories.append(0)
    else:
        calories[i] += int(snack)

print("Part 1: ", max(calories))

NUM_ELVES = 3
total = 0
for i in range(NUM_ELVES):
    total += int(calories.pop(calories.index(max(calories))))

print("Part 2: ", total)
